cmakebuild/docs/man/man3/m_getopt.3
.TH "m_getopt" 3 "Tue Feb 20 2018" "Mstdlib-1.0.0" \" -*- nroff -*-
.ad l
.nh
.SH NAME
m_getopt
.SH SYNOPSIS
.br
.PP
.SS "Typedefs"

.in +1c
.ti -1c
.RI "typedef struct M_getopt \fBM_getopt_t\fP"
.br
.ti -1c
.RI "typedef M_bool(* \fBM_getopt_nonopt_cb\fP) (size_t idx, const char *option, void *thunk)"
.br
.ti -1c
.RI "typedef M_bool(* \fBM_getopt_integer_cb\fP) (char short_opt, const char *long_opt, M_int64 *integer, void *thunk)"
.br
.ti -1c
.RI "typedef M_bool(* \fBM_getopt_decimal_cb\fP) (char short_opt, const char *long_opt, \fBM_decimal_t\fP *decimal, void *thunk)"
.br
.ti -1c
.RI "typedef M_bool(* \fBM_getopt_string_cb\fP) (char short_opt, const char *long_opt, const char *string, void *thunk)"
.br
.ti -1c
.RI "typedef M_bool(* \fBM_getopt_boolean_cb\fP) (char short_opt, const char *long_opt, M_bool boolean, void *thunk)"
.br
.in -1c
.SS "Enumerations"

.in +1c
.ti -1c
.RI "enum \fBM_getopt_error_t\fP { \fBM_GETOPT_ERROR_SUCCESS\fP = 0, \fBM_GETOPT_ERROR_INVALIDOPT\fP, \fBM_GETOPT_ERROR_INVALIDDATATYPE\fP, \fBM_GETOPT_ERROR_INVALIDORDER\fP, \fBM_GETOPT_ERROR_MISSINGVALUE\fP, \fBM_GETOPT_ERROR_NONOPTION\fP }"
.br
.in -1c
.SS "Functions"

.in +1c
.ti -1c
.RI "\fBM_getopt_t\fP * \fBM_getopt_create\fP (\fBM_getopt_nonopt_cb\fP cb)"
.br
.ti -1c
.RI "void \fBM_getopt_destroy\fP (\fBM_getopt_t\fP *g)"
.br
.ti -1c
.RI "char * \fBM_getopt_help\fP (const \fBM_getopt_t\fP *g)"
.br
.ti -1c
.RI "M_bool \fBM_getopt_addinteger\fP (\fBM_getopt_t\fP *g, char short_opt, const char *long_opt, M_bool val_required, const char *description, \fBM_getopt_integer_cb\fP cb)"
.br
.ti -1c
.RI "M_bool \fBM_getopt_adddecimal\fP (\fBM_getopt_t\fP *g, char short_opt, const char *long_opt, M_bool val_required, const char *description, \fBM_getopt_decimal_cb\fP cb)"
.br
.ti -1c
.RI "M_bool \fBM_getopt_addstring\fP (\fBM_getopt_t\fP *g, char short_opt, const char *long_opt, M_bool val_required, const char *description, \fBM_getopt_string_cb\fP cb)"
.br
.ti -1c
.RI "M_bool \fBM_getopt_addboolean\fP (\fBM_getopt_t\fP *g, char short_opt, const char *long_opt, M_bool val_required, const char *description, \fBM_getopt_boolean_cb\fP cb)"
.br
.ti -1c
.RI "\fBM_getopt_error_t\fP \fBM_getopt_parse\fP (\fBM_getopt_t\fP *g, const char *const *argv, int argc, const char **opt_fail, void *thunk)"
.br
.in -1c
.SH "Detailed Description"
.PP 
Command line argument parsing\&.
.PP
Handles parsing using a series of provided callbacks for each type\&. Auto conversion of the argument for the option type will take place\&. Callbacks are allowed to reject the option or argument\&. This will stop parsing\&.
.PP
If auto conversion is unwanted use the string option type\&. String options will alway shave their arguments passed unmodified\&.
.PP
Options can be marked as having required or optional arguments\&.
.PP
The boolean type is somewhat special\&. If marked as not having a required argument it is treated as a flag\&. For example -b would call the boolean callback with a value of M_TRUE\&. If a boolean is marked as val required then a value is required and the result of conversion (using M_str_istrue) will be passed to the callback\&.
.PP
For options that should \fInot\fP have an argument use the boolean type with val not required\&.
.PP
Supports auto generation of help message\&.
.PP
Valid characters for options (short/long) is all ASCII printable [!-~] except:
.IP "\(bu" 2
space
.IP "\(bu" 2
'-' (short or start/end long)
.IP "\(bu" 2
'='
.IP "\(bu" 2
'''
.IP "\(bu" 2
'''
.PP
.PP
To stop option processing and treat all following values as nonoptions use -- as an option\&.
.PP
An optional thunk can be passed in during parsing which will be passed to all callbacks\&. This can be used to collect all options into an object instead of stored in global variables\&.
.PP
Option callbacks will receive the short and long options associated with the option\&. If no short option was set the short_opt callback value will be 0\&. If no long option was set the long_opt callback value will be NULL\&.
.PP
Example:
.PP
.PP
.nf
static M_bool nonopt_cb(size_t idx, const char *option, void *thunk)
{
    M_printf("option='%s'\n", option);
}

static M_bool int_cb(char short_opt, const char *long_opt, M_int64 *integer, void *thunk)
{
    M_printf("short_opt='%d', long_opt='%s', integer='%lld'\n", short_opt, long_opt, integer);
}

int main(int argc, char **argv) {
    M_getopt_t *g; 
    char       *help;
    const char *fail;

    g = M_getopt_create(nonopt_cb);
    M_getopt_addinteger(g, 'i', "i1", M_TRUE, "DESCR 1", check_getopt_int_cb);

    help = M_getopt_help(g);
    M_printf("help=\n%s\n", help);
    M_free(help);

    if (M_getopt_parse(g, argv, argc, &fail, NULL) == M_GETOPT_ERROR_SUCCESS) {
        M_printf("Options parsed successfully\n");
    } else {
        M_printf("Options parse error: %s\n", fail);
    }

    M_getopt_destroy(g);
    return 0;
} 
.fi
.PP
 
.SH "Typedef Documentation"
.PP 
.SS "typedef struct M_getopt \fBM_getopt_t\fP"

.SS "typedef M_bool(* M_getopt_nonopt_cb) (size_t idx, const char *option, void *thunk)"
Callback for non-option parameters 
.SS "typedef M_bool(* M_getopt_integer_cb) (char short_opt, const char *long_opt, M_int64 *integer, void *thunk)"
Callback for integer data type 
.SS "typedef M_bool(* M_getopt_decimal_cb) (char short_opt, const char *long_opt, \fBM_decimal_t\fP *decimal, void *thunk)"
Callback for decimal data type 
.SS "typedef M_bool(* M_getopt_string_cb) (char short_opt, const char *long_opt, const char *string, void *thunk)"
Callback for string data type 
.SS "typedef M_bool(* M_getopt_boolean_cb) (char short_opt, const char *long_opt, M_bool boolean, void *thunk)"
Callback for boolean data type 
.SH "Enumeration Type Documentation"
.PP 
.SS "enum \fBM_getopt_error_t\fP"
Error codes\&. 
.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fIM_GETOPT_ERROR_SUCCESS \fP\fP
.TP
\fB\fIM_GETOPT_ERROR_INVALIDOPT \fP\fP
.TP
\fB\fIM_GETOPT_ERROR_INVALIDDATATYPE \fP\fP
.TP
\fB\fIM_GETOPT_ERROR_INVALIDORDER \fP\fP
.TP
\fB\fIM_GETOPT_ERROR_MISSINGVALUE \fP\fP
.TP
\fB\fIM_GETOPT_ERROR_NONOPTION \fP\fP
.SH "Function Documentation"
.PP 
.SS "\fBM_getopt_t\fP* M_getopt_create (\fBM_getopt_nonopt_cb\fP cb)"
Create a new getopt object\&.
.PP
\fBParameters:\fP
.RS 4
\fIcb\fP Callback to be called with non-option parameters\&. NULL if non-option parameters are not allowed\&.
.RE
.PP
\fBReturns:\fP
.RS 4
Getopt object\&. 
.RE
.PP

.SS "void M_getopt_destroy (\fBM_getopt_t\fP * g)"
Destroy a getopt object
.PP
\fBParameters:\fP
.RS 4
\fIg\fP Getopt object to destroy\&. 
.RE
.PP

.SS "char* M_getopt_help (const \fBM_getopt_t\fP * g)"
Output help text for command line options\&.
.PP
Components: <val> value is required [val] value is optional -s short option --long long option (type) Type such as integer, decimal \&.\&.\&. Type will not be printed for boolean options\&. Description Text description about the option
.PP
Example: 
.PP
.nf
-s \<val\> (type) Description 

.fi
.PP
 
.PP
.nf
--long [val] (type) 

.fi
.PP
 
.PP
.nf
-s, --long [val] (type) Description 

.fi
.PP
 
.PP
.nf
-s Description 

.fi
.PP
.PP
\fBParameters:\fP
.RS 4
\fIg\fP Getopt object\&. 
.RE
.PP

.SS "M_bool M_getopt_addinteger (\fBM_getopt_t\fP * g, char short_opt, const char * long_opt, M_bool val_required, const char * description, \fBM_getopt_integer_cb\fP cb)"
Add an integer parameter\&.
.PP
\fBParameters:\fP
.RS 4
\fIg\fP Getopt object 
.br
\fIshort_opt\fP Short option, must be alpha-numeric, case-sensitive\&. Pass 0 if not used 
.br
\fIlong_opt\fP Long option name, must be alpha-numeric or hyphens, case-insensitive\&. Can not start or end with hyphens\&. Pass NULL if not used\&. 
.br
\fIval_required\fP Whether or not the option requires a value\&. 
.br
\fIdescription\fP Field description\&. Used with output putting help message\&. 
.br
\fIcb\fP Callback to call with value\&. -1 will be passed if no value provided
.RE
.PP
\fBReturns:\fP
.RS 4
M_TRUE on success, M_FALSE on failure\&. 
.RE
.PP

.SS "M_bool M_getopt_adddecimal (\fBM_getopt_t\fP * g, char short_opt, const char * long_opt, M_bool val_required, const char * description, \fBM_getopt_decimal_cb\fP cb)"
Add a decimal parameter\&.
.PP
\fBParameters:\fP
.RS 4
\fIg\fP Getopt object 
.br
\fIshort_opt\fP Short option, must be alpha-numeric, case-sensitive\&. Pass 0 if not used 
.br
\fIlong_opt\fP Long option name, must be alpha-numeric or hyphens, case-insensitive\&. Can not start or end with hyphens\&. Pass NULL if not used\&. 
.br
\fIval_required\fP Whether or not the option requires a value\&. 
.br
\fIdescription\fP Field description\&. Used with output putting help message\&. 
.br
\fIcb\fP Callback to call with value\&. NULL will be passed if no value provided
.RE
.PP
\fBReturns:\fP
.RS 4
M_TRUE on success, M_FALSE on failure\&. 
.RE
.PP

.SS "M_bool M_getopt_addstring (\fBM_getopt_t\fP * g, char short_opt, const char * long_opt, M_bool val_required, const char * description, \fBM_getopt_string_cb\fP cb)"
Add a string parameter\&.
.PP
\fBParameters:\fP
.RS 4
\fIg\fP Getopt object 
.br
\fIshort_opt\fP Short option, must be alpha-numeric, case-sensitive\&. Pass 0 if not used 
.br
\fIlong_opt\fP Long option name, must be alpha-numeric or hyphens, case-insensitive\&. Can not start or end with hyphens\&. Pass NULL if not used\&. 
.br
\fIval_required\fP Whether or not the option requires a value\&. 
.br
\fIdescription\fP Field description\&. Used with output putting help message\&. 
.br
\fIcb\fP Callback to call with value\&. NULL will be passed if no value provided
.RE
.PP
\fBReturns:\fP
.RS 4
M_TRUE on success, M_FALSE on failure\&. 
.RE
.PP

.SS "M_bool M_getopt_addboolean (\fBM_getopt_t\fP * g, char short_opt, const char * long_opt, M_bool val_required, const char * description, \fBM_getopt_boolean_cb\fP cb)"
Add a boolean parameter\&.
.PP
\fBParameters:\fP
.RS 4
\fIg\fP Getopt object 
.br
\fIshort_opt\fP Short option, must be alpha-numeric, case-sensitive\&. Pass 0 if not used 
.br
\fIlong_opt\fP Long option name, must be alpha-numeric or hyphens, case-insensitive\&. Can not start or end with hyphens\&. Pass NULL if not used\&. 
.br
\fIval_required\fP Whether or not the option requires a value\&. If M_FALSE this is treated as a flag and will be treated as M_TRUE in the value of the callback\&. If M_FALSE a value cannot be provided\&. 
.br
\fIdescription\fP Field description\&. Used with output putting help message\&. 
.br
\fIcb\fP Callback to call with value\&. NULL will be passed if no value provided
.RE
.PP
\fBReturns:\fP
.RS 4
M_TRUE on success, M_FALSE on failure\&. 
.RE
.PP

.SS "\fBM_getopt_error_t\fP M_getopt_parse (\fBM_getopt_t\fP * g, const char *const * argv, int argc, const char ** opt_fail, void * thunk)"
Parse command line arguments\&.
.PP
\fBParameters:\fP
.RS 4
\fIg\fP Getopt object 
.br
\fIargv\fP Array of arguments\&. Will not be modified\&. 
.br
\fIargc\fP Number of arguments\&. 
.br
\fIopt_fail\fP On failure will have the argument that caused the failure\&. 
.br
\fIthunk\fP Thunk that will be passed to callbacks\&.
.RE
.PP
\fBReturns:\fP
.RS 4
Result\&. 
.RE
.PP

.SH "Author"
.PP 
Generated automatically by Doxygen for Mstdlib-1\&.0\&.0 from the source code\&.


base/data/m_getopt.c
/* The MIT License (MIT)
 * 
 * Copyright (c) 2015 Main Street Softworks, Inc.
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#include "m_config.h"

#include <mstdlib/mstdlib.h>
#include "data/m_getopt_int.h"

/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */

static M_bool M_getopt_validate_short_opt(char short_opt)
{
	if (short_opt == 0)
		return M_TRUE;

	if (!M_chr_isgraph(short_opt))
		return M_FALSE;

	if (short_opt == '-' || short_opt == '=' || short_opt == '"' || short_opt == '\'')
		return M_FALSE;

	return M_TRUE;
}

static M_bool M_getopt_validate_long_opt(const char *long_opt)
{
	if (long_opt == NULL)
		return M_TRUE;

	if (*long_opt == '\0')
		return M_FALSE;

	if (!M_str_isgraph(long_opt))
		return M_FALSE;

	if (*long_opt == '-' || long_opt[M_str_len(long_opt)-1] == '-')
		return M_FALSE;

	if (M_str_chr(long_opt, '=') != NULL || M_str_chr(long_opt, '"') != NULL || M_str_chr(long_opt, '\'') != NULL)
		return M_FALSE;

	return M_TRUE;
}

static M_getopt_option_t *M_getopt_create_option(M_getopt_type_t type, char short_opt, const char *long_opt, M_bool val_required, const char *description)
{
	M_getopt_option_t *opt;

	/* Validate we have a short, long or both and they are valid. */
	if ((short_opt == 0 && long_opt == NULL) ||
		!M_getopt_validate_short_opt(short_opt) || !M_getopt_validate_long_opt(long_opt))
	{
		return NULL;
	}

	opt               = M_malloc_zero(sizeof(*opt));
	opt->type         = type;
	opt->short_opt    = short_opt;
	opt->long_opt     = long_opt!=NULL?M_strdup(long_opt):NULL;
	opt->val_required = val_required;
	opt->description  = M_strdup(description);

	return opt;
}

static void M_getopt_destroy_option(M_getopt_option_t *opt)
{
	if (opt == NULL)
		return;

	M_free(opt->long_opt);
	M_free(opt->description);

	M_free(opt);
}

static M_bool M_getopt_append_option(M_getopt_t *g, M_getopt_option_t *opt)
{
	if (g == NULL || opt == NULL)
		return M_FALSE;;

	/* Check if this option has already been added. */
	if ((opt->short_opt != 0 && M_hash_u64vp_get(g->short_opts, (M_uint64)opt->short_opt, NULL)) || (g->long_opts != NULL && M_hash_strvp_get(g->long_opts, opt->long_opt, NULL)))
		return M_FALSE;

	if (opt->short_opt != 0)
		M_hash_u64vp_insert(g->short_opts, (M_uint64)opt->short_opt, opt);
	if (opt->long_opt != NULL)
		M_hash_strvp_insert(g->long_opts, opt->long_opt, opt);

	if (g->last == NULL) {
		g->first      = opt;
		g->last       = opt;
	} else {
		g->last->next = opt;
		g->last       = opt;
	}

	return M_TRUE;
}

/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */

M_getopt_t *M_getopt_create(M_getopt_nonopt_cb cb)
{
	M_getopt_t *g;

	g = M_malloc_zero(sizeof(*g));
	g->short_opts  = M_hash_u64vp_create(8, 75, M_HASH_U64VP_NONE, NULL);
	g->long_opts   = M_hash_strvp_create(8, 75, M_HASH_STRVP_CASECMP, NULL);
	g->nonopt_cb   = cb;

	return g;
}

void M_getopt_destroy(M_getopt_t *g)
{
	M_getopt_option_t *opt;
	M_getopt_option_t *next;

	if (g == NULL)
		return;

	M_hash_u64vp_destroy(g->short_opts, M_FALSE);
	M_hash_strvp_destroy(g->long_opts, M_FALSE);

	opt = g->first;
	while (opt != NULL) {
		/* Store the next option because the option will be destroyed. */
		next = opt->next;
		M_getopt_destroy_option(opt);
		opt = next;
	}

	M_free(g);
}

/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */

char *M_getopt_help(const M_getopt_t *g)
{
	M_buf_t           *buf;
	M_getopt_option_t *opt;
	M_bool             has_short;

	if (g == NULL || g->first == NULL)
		return NULL;

	buf = M_buf_create();

	opt = g->first;
	while (opt != NULL) {
		has_short = M_FALSE;

		M_buf_add_fill(buf, ' ', 2);

		if (opt->short_opt != 0) {
			has_short = M_TRUE;
			M_buf_add_byte(buf, '-');
			M_buf_add_char(buf, opt->short_opt);
		}
		if (opt->long_opt != NULL) {
			if (has_short) {
				M_buf_add_str(buf, ", ");
			}
			M_buf_add_str(buf, "--");
			M_buf_add_str(buf, opt->long_opt);
		}

		if (opt->type != M_GETOPT_TYPE_BOOLEAN || (opt->type == M_GETOPT_TYPE_BOOLEAN && opt->val_required)) {
			M_buf_add_byte(buf, ' ');
			if (opt->val_required) {
				M_buf_add_str(buf, "<val>");
			} else {
				M_buf_add_str(buf, "[val]");
			}

			M_buf_add_byte(buf, ' ');

			M_buf_add_byte(buf, '(');
			switch (opt->type) {
				case M_GETOPT_TYPE_INTEGER:
					M_buf_add_str(buf, "integer");
					break;
				case M_GETOPT_TYPE_DECIMAL:
					M_buf_add_str(buf, "decimal");
					break;
				case M_GETOPT_TYPE_STRING:
					M_buf_add_str(buf, "string");
					break;
				case M_GETOPT_TYPE_BOOLEAN:
					M_buf_add_str(buf, "boolean");
					break;
				case M_GETOPT_TYPE_UNKNOWN:
					break;
			}
			M_buf_add_byte(buf, ')');
		}

		if (opt->description != NULL) {
			M_buf_add_byte(buf, ' ');
			M_buf_add_str(buf, opt->description);
		}

		M_buf_add_byte(buf, '\n');
		opt = opt->next;
	}

	return M_buf_finish_str(buf, NULL);
}

/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */

M_bool M_getopt_addinteger(M_getopt_t *g, char short_opt, const char *long_opt, M_bool val_required, const char *description, M_getopt_integer_cb cb)
{
	M_getopt_option_t *opt;

	if (g == NULL || cb == NULL)
		return M_FALSE;

	opt = M_getopt_create_option(M_GETOPT_TYPE_INTEGER, short_opt, long_opt, val_required, description);
	if (opt == NULL)
		return M_FALSE;
	opt->cb.integer_cb = cb;

	if (!M_getopt_append_option(g, opt)) {
		M_getopt_destroy_option(opt);
		return M_FALSE;
	}
	return M_TRUE;
}

M_bool M_getopt_adddecimal(M_getopt_t *g, char short_opt, const char *long_opt, M_bool val_required, const char *description, M_getopt_decimal_cb cb)
{
	M_getopt_option_t *opt;

	if (g == NULL || cb == NULL)
		return M_FALSE;

	opt = M_getopt_create_option(M_GETOPT_TYPE_DECIMAL, short_opt, long_opt, val_required, description);
	if (opt == NULL)
		return M_FALSE;
	opt->cb.decimal_cb = cb;

	if (!M_getopt_append_option(g, opt)) {
		M_getopt_destroy_option(opt);
		return M_FALSE;
	}
	return M_TRUE;
}

M_bool M_getopt_addstring(M_getopt_t *g, char short_opt, const char *long_opt, M_bool val_required, const char *description, M_getopt_string_cb cb)
{
	M_getopt_option_t *opt;

	if (g == NULL || cb == NULL)
		return M_FALSE;

	opt = M_getopt_create_option(M_GETOPT_TYPE_STRING, short_opt, long_opt, val_required, description);
	if (opt == NULL)
		return M_FALSE;
	opt->cb.string_cb = cb;

	if (!M_getopt_append_option(g, opt)) {
		M_getopt_destroy_option(opt);
		return M_FALSE;
	}
	return M_TRUE;
}

M_bool M_getopt_addboolean(M_getopt_t *g, char short_opt, const char *long_opt, M_bool val_required, const char *description, M_getopt_boolean_cb cb)
{
	M_getopt_option_t *opt;

	if (g == NULL || cb == NULL)
		return M_FALSE;

	opt = M_getopt_create_option(M_GETOPT_TYPE_BOOLEAN, short_opt, long_opt, val_required, description);
	if (opt == NULL)
		return M_FALSE;
	opt->cb.boolean_cb = cb;

	if (!M_getopt_append_option(g, opt)) {
		M_getopt_destroy_option(opt);
		return M_FALSE;
	}
	return M_TRUE;
}
